File: CMPS12B/HW3/Balanced.java
/*
*Balanced.java
*Juan Ayala
*jumaayal
*This program checks if a string of characters is balanced with brackets
*/
import java.util.*;
import java.io.*;
import java.util.Scanner;
class Balanced extends StackLs{
  public static void main(String[] args)throws IOException {
    if(args.length < 2){
      System.out.println("Usage: java –jar FileCopy.jar <input file> <outputfile>");
      System.exit(1);
    }
    // open files
    Scanner in = new Scanner(new File(args[0]));
    PrintWriter out = new PrintWriter(new FileWriter(args[1]));
    StackLs stack = new StackLs(); //Creates a stack
    //Will continue operating till there is no line of characters
    while(in.hasNextLine()){
      String line = in.nextLine(); //Puts character string into line
      int stringLength= line.length();

      //Goes through each character and if a bracket then it pushes or pops into stack
      for(int i=0; i<stringLength;i++){
        String symbol = Character.toString(line.charAt(i));

      //if it is an opening bracket it will be pushed in
        if(symbol.equals("(") || symbol.equals("[")|| symbol.equals("<")|| symbol.equals("{")){
          stack.push(symbol);
        }
      //If it is the closing bracket it should pop out the opening bracket corresponding to it
        if(symbol.equals("}") || symbol.equals(")")|| symbol.equals(">")||symbol.equals("]")){
          String other = otherBracket(symbol);

        //If it is a closing bracket with no opening bracket then push bracket
          if( !stack.isEmpty() && stack.peek().equals(other)){
            String poppedItem =stack.pop();
          }else{
            stack.push(symbol);
          }
        }
      }
      /*
      *According to method above, if it was ballanced the stack would be empty
      *due to the right amount of open brackets being pushed and popped with
      *corresponding closing brackets
      */
      if(!stack.isEmpty()){
        out.println("N");
      }else{
        out.println("Y");
      }
      stack.emptyOut(); //Emptys out the stack to be used for new string of characters
    }
    in.close();
    out.close();
  }

//This checks whether the next popped item is the opening bracket of the closing bracket
  public static String otherBracket (String item){
    if(item.equals(")")){
      return "(";
    }
    if(item.equals("]")){
      return "[";
    }
    if(item.equals(">")){
      return "<";
    }
    if(item.equals("}")){
      return "{";
    }
    return "";
  }
}

//This class is for the stack methods
class StackLs{
  Node top; //uses the node class or the linked list
  public StackLs(){
    top=null;
  }
  //Checkes whether stack is empty
  public boolean isEmpty(){
    return top == null;
  }
  //Pushes an item to the stack
  public void push(String symbol){
    top = new Node(symbol,top);
  }
  //takes off the recent item makes the next thing the top
  public String pop(){
    if(!isEmpty()){
      Node holder = top;
      top=top.getNext();
      return holder.getItem();
    }else{
      throw new EmptyStackException();
    }
  }
  //Emptys out the stack to create a new stack
  public void emptyOut(){
    top=null;
  }
  //Shows what the next item is without impacting the stack
  public String peek() throws EmptyStackException{
    if(!isEmpty()){
      return top.getItem();
    }else{
      throw new EmptyStackException();
    }
  }
}

//This create a class that can hold a string and its own class
//Used to create a refenced stack
class Node{
  Node next;
  String item;

  //holds a string and Node
  //Takes an item and puts it in a node
  Node(String newItem){
    item=newItem;
    next=null;
  }

  //Places a node and string into current node
  Node(String newItem, Node nextNode){
    item = newItem;
    next=nextNode;
  }

  //places a string in current node
  void putItem(String newItem){
    item=newItem;
  }
  //Returns item in current Node
  String getItem(){
    return item;
  }
  //Places a node in current node
  void putNext(Node newNext){
    next=newNext;
  }
  //Returns the node within the current node
  Node getNext(){
    return next;
  }
}

File: CMPS12B/Lab4/balanced.c
/*
*balanced.c
*Juan Ayala
*jumaayal
*Uses array based stack to see if text is balanced
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TRUE 1
#define FALSE 0
typedef struct{
  int top;
  int max;
  char brackets[80];
}stack;
char peek(stack * list);
void push(stack * list,char symbol);
char pop(stack * list);
int isEmpty(stack * list);
void popAll(stack *list);
char otherBracket(char item);
int main(int argc, char* argv[]){
  FILE* in; /*file handle for input*/
  FILE* out;/*file handle for output*/
  char input[80];


  /* check command line for correct number of arguments*/
  if(argc != 3){
    printf("usage: %s <input file> <output file>\n",argv[0]);
    exit(EXIT_FAILURE);
  }

  /* open input file for reading*/
  in = fopen(argv[1], "r");
  if( in==NULL ){
    printf("Unable to read from file %s\n", argv[1]);
    exit(EXIT_FAILURE);
  }

  /* open output file for writing */
  out = fopen(argv[2], "w");
  if( out==NULL ){
    printf("Unable to write to file %s\n", argv[2]);
    exit(EXIT_FAILURE);
  }
  /*Setting up the stack*/
  stack load;
  load.max=80;
  load.top=-1;

  /*Will continue running till there is no input*/
  while(fgets(input,80,in)!=NULL){
    int inLength= strlen(input);
    int i;

    //Runs in a loop until last character
    for(i=0;i<inLength;i++){

      //If it is an opening character push into stack
      if(input[i]== '(' || input[i]== '[' || input[i]== '<' || input[i]=='{'){
        push(&load, input[i]);
      }

      /*
      *if it is a closing bracket, check to see if stack has corresponding opening bracket
      *Pop if stack has corresponding brackets
      *Push if not
      */
      if(input[i]== ')' || input[i]== ']' || input[i]== '>' || input[i]=='}'){
        char other = otherBracket(input[i]);
        if(!isEmpty(&load) && peek(&load)==other){
          pop(&load);
        }else{
          push(&load, input[i]);
        }
      }
    }

    /*If there is nothing in stack then it is balanced*/
    if(!isEmpty(&load)){
      fprintf(out, "N\n");
    }else{
      fprintf(out, "Y\n");
    }
    popAll(&load); //pop all so stack can be reused
  }
  fclose(in);
  fclose(out);
  return EXIT_SUCCESS;
}

//adds item to next array index and changes top to be new index
void push(stack * list,char symbol){
  list->top++;
  list->brackets[list->top]=symbol;
}

//Returns the popped item and lowers top index
char pop(stack * list){
  char holder = (list->top);
  list->top--;
  return holder;
}

//Checks if top is at -1 meaning it is empty
int isEmpty(stack * list){
  if ((*list).top <= -1){
    return TRUE;
  }else{
    return FALSE;
  }
}

//Changes all values to zero from top and lowers top to empty
void popAll(stack *list){
  int i;
  for(i=0; i<=(list->max);i++){
    list->brackets[i] =0;
  }
  list->top = -1;
}

//Returns the item at the top
char peek(stack * list){
  return list->brackets[list->top];
}

//Returns what the opening bracket should be compared to closing
char otherBracket(char item){
  if(item==')'){
    return '(';
  }
  if(item==']'){
    return '[';
  }
  if(item=='>'){
    return '<';
  }
  if(item=='}'){
    return '{';
  }
  return '0';
}
